academic-services/knowledge-exploration-service/tutorial-grammar-design.md

```diff
@@ -0,0 +1,392 @@
+
+To enable fuzzy search or keyword search, we may also transform the attributes before indexing the attributes. For example, we enable abstract keywords search by transforming the abstract of a publication `Abstract.OriginalName` into `Abstract.Words` such that unique words from the abstract can be indexed.
+
+## Design a search grammar
+
+Now that we have the appropriate data fields in our index, we will design a grammar tailored to process natural language search queries. In this tutorial, we will use the [sample grammar](samplePrivateLibraryData.linked.search.grammar.xml) to illustrate how to support **direct attribute search**, **multiple attribute search**, **partial attribute search**, **semantic search**, and **drop terms with penalties**.
+
+### Support direct attribute search
+
+Direct attribute search is used for processing queries that only contain a single publication attribute. For example, a user may want to look up a publication by its title, doi, or full text url. These are basically lookup queries using attributes that can be identifiers of the entities themselves. To craft a grammar for direct attribute search, we can use the `<attrref>` element:
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_direct_attribute_search":::
+
+This allows us to process queries such as "paper titled An Overview of Microsoft Academic Service (MAS) and Applications" or "paper with doi 10.1145/2740908.2742839"
+
+We also need to return the corresponding KES query expression once the attribute is matched. The `<attrref uri="libraryPapers#Title.Name" name="q"/>` element will store the matched paper title in a variable called "q". We can then craft our return query by using the following:
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_create_return_query":::
+
+### Support multiple attribute search
+
+Multiple attribute search is used for processing exploratory search queries that contain multiple attribute terms. For example, a user may want to search for publications given a few fields of study and affiliations.
+
+We use the repeat item `<item repeat="1-INF" repeat-logprob="-1">` element to process queries that contains multiple attributes by continously matching remaining query terms using different search grammars.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_multiple_attribute_search_loop":::
+
+This allows us to process queries such as "papers from microsoft about machine learning"
+
+### Support partial attribute search
```

Comment:
>Support partial attribute search [](start = 4, length = 32)

"partial attribute queries" should refer to prefix matching, which you have not enabled.

What you're describing here is good old fashioned keyword search.

Reply:
Changed to keyword queries?


---
In reply to: [533874968](https://github.com/MicrosoftDocs/microsoft-academic-services/pull/194#discussion_r533874968) [](ancestors = 533874968)